File: packages/asdl-slots/src/asdl_slots/lifecycle/gc.py
"""Garbage collection: free slots whose PRs have merged or closed."""

from __future__ import annotations

import dataclasses
from collections.abc import Sequence
from dataclasses import dataclass

from asdl_core.gh.types import PRGatewayFailure, PRLookupMiss, PRSummary
from asdl_slots.context import SlotsCliContext
from asdl_slots.inventory import SlotRecord, build_slot_inventory
from asdl_slots.lifecycle.operation_state import (
    operation_in_progress_detail,
    slot_operation_in_progress_message,
)
from asdl_slots.lifecycle.outcomes import (
    FreedSlot,
    SlotFreeCleanupAction,
    SlotFreeCleanupResult,
    SlotGcAction,
    SlotGcEntry,
    SlotGcOutcome,
    SlotGcPlan,
    SlotLifecycleFailure,
)
from asdl_slots.lifecycle.release_cleanup import (
    execute_release_cleanup,
    plan_release_cleanup,
)
from asdl_slots.lifecycle.release_target import (
    ReleaseTargetFailure,
    release_assigned_slot_target,
)


@dataclass(frozen=True)
class _GcCounts:
    freed_count: int
    kept_count: int
    skipped_count: int
    error_count: int


def plan_gc(slots_ctx: SlotsCliContext) -> SlotGcPlan | SlotLifecycleFailure:
    """Classify assigned slots for garbage collection without mutating state."""
    inventory = build_slot_inventory(
        slots_ctx.git,
        main_repo_root=slots_ctx.repo.main_repo_root,
    )
    if inventory.pool_size == 0:
        return _gc_pool_empty_failure()

    entries: list[SlotGcEntry] = []
    would_free_count = 0

    for record in inventory.records:
        if record.branch is None:
            continue
        if record.operation is not None:
            entries.append(
                _entry_from_record(
                    record,
                    "skipped_operation",
                    message=operation_in_progress_detail(
                        branch_name=record.branch,
                        worktree_path=record.path,
                        operation=record.operation,
                        action="running slot gc",
                    ),
                )
            )
            continue
        pr_result = slots_ctx.pr.get_pr_for_branch(record.branch)

        if isinstance(pr_result, PRLookupMiss):
            entries.append(_entry_from_record(record, "kept_no_pr"))
            continue

        if isinstance(pr_result, PRGatewayFailure):
            entries.append(
                _entry_from_record(
                    record,
                    "error",
                    message=(
                        pr_result.stderr
                        or pr_result.stdout
                        or f"gh pr view exited {pr_result.returncode}"
                    ),
                )
            )
            continue

        if pr_result.state == "OPEN":
            entries.append(_entry_from_record(record, "kept_open_pr", pr_result=pr_result))
            continue

        entries.append(_entry_from_record(record, "would_free", pr_result=pr_result))
        would_free_count += 1

    return SlotGcPlan(entries=tuple(entries), would_free_count=would_free_count)


def plan_gc_cleanup(
    slots_ctx: SlotsCliContext,
    plan: SlotGcPlan,
    cleanup_actions: Sequence[SlotFreeCleanupAction],
) -> tuple[SlotFreeCleanupResult, ...]:
    """Plan GC cleanup without freeing slots or deleting local branches."""
    targets = _gc_free_targets(plan.entries)
    if not targets or not cleanup_actions:
        return ()
    trunk_branch = slots_ctx.git.get_trunk_branch()
    return plan_release_cleanup(slots_ctx, targets, cleanup_actions, trunk_branch=trunk_branch)


def execute_gc_plan(
    slots_ctx: SlotsCliContext,
    plan: SlotGcPlan,
    *,
    cleanup_actions: Sequence[SlotFreeCleanupAction] = (),
) -> SlotGcOutcome:
    """Free every ``would_free`` entry in ``plan``; pass through the rest."""
    inventory = build_slot_inventory(
        slots_ctx.git,
        main_repo_root=slots_ctx.repo.main_repo_root,
    )
    trunk = slots_ctx.git.get_trunk_branch()
    entries: list[SlotGcEntry] = []
    freed_entries: list[SlotGcEntry] = []

    for entry in plan.entries:
        if entry.action != "would_free":
            entries.append(entry)
            continue

        target = _freed_slot_from_gc_entry(entry)
        result = release_assigned_slot_target(
            slots_ctx.git,
            inventory,
            target,
            trunk,
        )
        if isinstance(result, ReleaseTargetFailure):
            entries.append(_entry_from_release_failure(entry, result))
            continue

        freed_entry = _with_action(entry, "freed")
        entries.append(freed_entry)
        freed_entries.append(freed_entry)

    if cleanup_actions and freed_entries:
        cleanup = execute_release_cleanup(
            slots_ctx,
            _gc_free_targets(freed_entries),
            cleanup_actions,
            trunk_branch=trunk,
        )
        entries = list(_with_cleanup_by_slot(entries, cleanup))

    counts = _count_gc_actions(entries)
    return SlotGcOutcome(
        entries=tuple(entries),
        freed_count=counts.freed_count,
        kept_count=counts.kept_count,
        skipped_count=counts.skipped_count,
        error_count=counts.error_count,
        dry_run=False,
        cleanup_error_count=_gc_cleanup_error_count(entries),
    )


def outcome_from_gc_plan(
    plan: SlotGcPlan,
    *,
    dry_run: bool,
    cleanup: Sequence[SlotFreeCleanupResult] = (),
) -> SlotGcOutcome:
    """Turn a GC plan and precomputed cleanup results into a renderable outcome."""
    entries = _with_cleanup_by_slot(plan.entries, cleanup) if cleanup else plan.entries
    counts = _count_gc_actions(entries)
    return SlotGcOutcome(
        entries=entries,
        freed_count=counts.freed_count,
        kept_count=counts.kept_count,
        skipped_count=counts.skipped_count,
        error_count=counts.error_count,
        dry_run=dry_run,
        cleanup_error_count=_gc_cleanup_error_count(entries),
    )


def garbage_collect_slots(
    slots_ctx: SlotsCliContext,
    *,
    dry_run: bool,
) -> SlotGcOutcome | SlotLifecycleFailure:
    """Plan the GC sweep and execute it unless ``dry_run`` is true."""
    plan = plan_gc(slots_ctx)
    if isinstance(plan, SlotLifecycleFailure):
        return plan
    if dry_run:
        return outcome_from_gc_plan(plan, dry_run=True)
    return execute_gc_plan(slots_ctx, plan)


def _gc_pool_empty_failure() -> SlotLifecycleFailure:
    return SlotLifecycleFailure(
        error_type="pool_empty",
        message="No managed slots configured. Run `slot init --size N` first.",
    )


def _entry_from_record(
    record: SlotRecord,
    action: SlotGcAction,
    *,
    pr_result: PRSummary | None = None,
    message: str | None = None,
) -> SlotGcEntry:
    assert record.branch is not None
    return SlotGcEntry(
        slot_name=record.slot_name,
        branch_name=record.branch,
        worktree_path=record.path,
        action=action,
        pr_number=pr_result.number if pr_result is not None else None,
        pr_state=pr_result.state if pr_result is not None else None,
        pr_url=pr_result.url if pr_result is not None else None,
        message=message,
    )


def _entry_from_release_failure(
    entry: SlotGcEntry,
    failure: ReleaseTargetFailure,
) -> SlotGcEntry:
    if failure.reason == "slot_not_assigned":
        return _with_action(
            entry,
            "error",
            message=(
                f"slot {entry.slot_name} was not assigned to {entry.branch_name} "
                "during free (state changed between plan and execute)."
            ),
        )
    if failure.reason == "operation_in_progress":
        return _with_action(
            entry,
            "skipped_operation",
            message=slot_operation_in_progress_message(
                slot_name=failure.slot_name,
                branch_name=failure.branch_name,
                worktree_path=failure.worktree_path,
                operation=failure.operation or "operation",
                action="running slot gc",
            ),
        )
    if failure.reason == "dirty_worktree":
        return _with_action(
            entry,
            "skipped_dirty",
            message=f"worktree has uncommitted changes at {failure.worktree_path}",
        )
    return _with_action(entry, "error", message=_detach_failure_message(failure))


def _detach_failure_message(failure: ReleaseTargetFailure) -> str:
    detach_ref = failure.detach_ref or "target ref"
    detach_error = f": {failure.detach_error}" if failure.detach_error else ""
    return (
        f"Failed to detach {failure.slot_name} at {failure.worktree_path} "
        f"to {detach_ref}{detach_error}"
    )


def _with_action(
    entry: SlotGcEntry,
    action: SlotGcAction,
    *,
    message: str | None = None,
) -> SlotGcEntry:
    return dataclasses.replace(entry, action=action, message=message)


def _with_cleanup(
    entry: SlotGcEntry,
    cleanup: Sequence[SlotFreeCleanupResult],
) -> SlotGcEntry:
    return dataclasses.replace(entry, cleanup=tuple(cleanup))


def _freed_slot_from_gc_entry(entry: SlotGcEntry) -> FreedSlot:
    return FreedSlot(
        slot_name=entry.slot_name,
        branch_name=entry.branch_name,
        worktree_path=entry.worktree_path,
    )


def _gc_free_targets(entries: Sequence[SlotGcEntry]) -> tuple[FreedSlot, ...]:
    return tuple(
        _freed_slot_from_gc_entry(entry)
        for entry in entries
        if entry.action in ("would_free", "freed")
    )


def _with_cleanup_by_slot(
    entries: Sequence[SlotGcEntry],
    cleanup: Sequence[SlotFreeCleanupResult],
) -> tuple[SlotGcEntry, ...]:
    cleanup_by_target: dict[tuple[str, str], list[SlotFreeCleanupResult]] = {}
    for result in cleanup:
        cleanup_by_target.setdefault((result.slot_name, result.branch_name), []).append(result)
    return tuple(
        _with_cleanup(entry, cleanup_by_target.get((entry.slot_name, entry.branch_name), ()))
        for entry in entries
    )


def _gc_cleanup_error_count(entries: Sequence[SlotGcEntry]) -> int:
    return sum(1 for entry in entries for cleanup in entry.cleanup if cleanup.status == "error")


def _count_gc_actions(entries: Sequence[SlotGcEntry]) -> _GcCounts:
    freed = 0
    kept = 0
    skipped = 0
    error = 0
    for entry in entries:
        if entry.action in ("freed", "would_free"):
            freed += 1
        elif entry.action in ("kept_open_pr", "kept_no_pr"):
            kept += 1
        elif entry.action in ("skipped_dirty", "skipped_operation"):
            skipped += 1
        elif entry.action == "error":
            error += 1
    return _GcCounts(
        freed_count=freed,
        kept_count=kept,
        skipped_count=skipped,
        error_count=error,
    )

File: ts/packages/capabilities/slot/src/lifecycle/gc.ts
import type { RepoSlotContext } from "../context.ts";
import type { PrState, PrSummary } from "../gateways/pr.ts";
import { prFailureMessage } from "../gateways/pr.ts";
import { buildSlotInventory, type SlotRecord } from "../inventory.ts";
import { slotOperationMessage, type LifecycleResult } from "./common.ts";
import {
	executeReleaseCleanup,
	planReleaseCleanup,
	type SlotFreeCleanupAction,
	type SlotFreeCleanupResult,
} from "./release-cleanup.ts";
import { detachFailureMessage } from "./free.ts";
import {
	releaseAssignedSlotTarget,
	type FreedSlot,
	type ReleaseTargetFailure,
} from "./release-target.ts";

export type SlotGcAction =
	| "freed"
	| "would-free"
	| "kept-open-pr"
	| "kept-no-pr"
	| "skipped-dirty"
	| "skipped-operation"
	| "error";

export interface SlotGcEntry {
	slotName: string;
	branchName: string;
	worktreePath: string;
	action: SlotGcAction;
	prNumber: number | null;
	prState: PrState | null;
	prUrl: string | null;
	message: string | null;
	cleanup: readonly SlotFreeCleanupResult[];
}

export interface SlotGcPlan {
	entries: readonly SlotGcEntry[];
	wouldFreeCount: number;
}

export interface SlotGcOutcome {
	entries: readonly SlotGcEntry[];
	freedCount: number;
	keptCount: number;
	skippedCount: number;
	errorCount: number;
	dryRun: boolean;
	cleanupErrorCount: number;
}

export async function planGc(ctx: RepoSlotContext): Promise<LifecycleResult<SlotGcPlan>> {
	const inventory = await buildSlotInventory(ctx.git, { mainRepoRoot: ctx.repo.mainRepoRoot });
	if (inventory.records.length === 0)
		return {
			type: "failure",
			failure: {
				errorType: "pool-empty",
				message: "No managed slots configured. Run `slot init --size N` first.",
			},
		};
	const prRecords = inventory.records.filter(
		(record) => record.branch !== null && record.operation === null,
	);
	const prBranches = prRecords.flatMap((record) => (record.branch === null ? [] : [record.branch]));
	const prLookup = await ctx.pr.getPrsForBranches(prBranches);
	if (prLookup.type === "failure")
		return {
			type: "failure",
			failure: {
				errorType: "pr-lookup-failed",
				message: prFailureMessage(prLookup.failure, "gh api graphql exited"),
			},
		};
	const entries: SlotGcEntry[] = [];
	let wouldFreeCount = 0;
	for (const record of inventory.records) {
		if (record.branch === null) continue;
		if (record.operation !== null) {
			entries.push(
				entryFromRecord(record, "skipped-operation", {
					message: slotOperationMessage(record, { action: "running slot gc" }),
				}),
			);
			continue;
		}
		const lookup = prLookup.resultsByBranch.get(record.branch);
		if (lookup === undefined)
			return {
				type: "failure",
				failure: {
					errorType: "pr-lookup-failed",
					message: `PR lookup did not return a result for ${record.branch}`,
				},
			};
		if (lookup.type === "miss") {
			entries.push(entryFromRecord(record, "kept-no-pr"));
			continue;
		}
		if (lookup.type === "failure")
			return {
				type: "failure",
				failure: {
					errorType: "pr-lookup-failed",
					message: prFailureMessage(lookup.failure, "gh api graphql exited"),
				},
			};
		if (lookup.pr.state === "OPEN") {
			entries.push(entryFromRecord(record, "kept-open-pr", { pr: lookup.pr }));
			continue;
		}
		entries.push(entryFromRecord(record, "would-free", { pr: lookup.pr }));
		wouldFreeCount += 1;
	}
	return { type: "ok", outcome: { entries, wouldFreeCount: wouldFreeCount } };
}

export async function planGcCleanup(
	ctx: RepoSlotContext,
	plan: SlotGcPlan,
	cleanupActions: readonly SlotFreeCleanupAction[],
): Promise<readonly SlotFreeCleanupResult[]> {
	const targets = gcFreeTargets(plan.entries);
	if (targets.length === 0 || cleanupActions.length === 0) return [];
	return await planReleaseCleanup({
		ctx,
		targets,
		cleanupActions,
		trunkBranch: await ctx.git.getTrunkBranch(),
	});
}

export async function executeGcPlan(
	ctx: RepoSlotContext,
	plan: SlotGcPlan,
	options: { cleanupActions?: readonly SlotFreeCleanupAction[] | undefined } = {},
): Promise<SlotGcOutcome> {
	const inventory = await buildSlotInventory(ctx.git, { mainRepoRoot: ctx.repo.mainRepoRoot });
	const trunk = await ctx.git.getTrunkBranch();
	let entries: SlotGcEntry[] = [];
	const freedEntries: SlotGcEntry[] = [];
	for (const entry of plan.entries) {
		if (entry.action !== "would-free") {
			entries.push(entry);
			continue;
		}
		const result = await releaseAssignedSlotTarget({
			git: ctx.git,
			inventory,
			target: freedSlotFromGcEntry(entry),
			trunkBranch: trunk,
		});
		if ("reason" in result) {
			entries.push(entryFromReleaseFailure(entry, result));
			continue;
		}
		const freed = withAction(entry, "freed");
		entries.push(freed);
		freedEntries.push(freed);
	}
	const cleanupActions = options.cleanupActions ?? [];
	if (cleanupActions.length > 0 && freedEntries.length > 0) {
		const cleanup = await executeReleaseCleanup({
			ctx,
			targets: gcFreeTargets(freedEntries),
			cleanupActions,
			trunkBranch: trunk,
		});
		entries = withCleanupBySlot(entries, cleanup);
	}
	return outcomeFromEntries(entries, false);
}

export function outcomeFromGcPlan(
	plan: SlotGcPlan,
	options: { isDryRun: boolean; cleanup?: readonly SlotFreeCleanupResult[] | undefined },
): SlotGcOutcome {
	const entries =
		options.cleanup === undefined || options.cleanup.length === 0
			? plan.entries
			: withCleanupBySlot(plan.entries, options.cleanup);
	return outcomeFromEntries(entries, options.isDryRun);
}

function entryFromRecord(
	record: SlotRecord,
	action: SlotGcAction,
	options: { pr?: PrSummary | undefined; message?: string | undefined } = {},
): SlotGcEntry {
	if (record.branch === null) throw new Error(`gc record ${record.slotName} is not assigned`);
	return {
		slotName: record.slotName,
		branchName: record.branch,
		worktreePath: record.path,
		action,
		prNumber: options.pr?.number ?? null,
		prState: options.pr?.state ?? null,
		prUrl: options.pr?.url ?? null,
		message: options.message ?? null,
		cleanup: [],
	};
}

function entryFromReleaseFailure(entry: SlotGcEntry, failure: ReleaseTargetFailure): SlotGcEntry {
	if (failure.reason === "slot-not-assigned")
		return withAction(
			entry,
			"error",
			`slot ${entry.slotName} was not assigned to ${entry.branchName} during free (state changed between plan and execute).`,
		);
	if (failure.reason === "operation-in-progress")
		return withAction(
			entry,
			"skipped-operation",
			`${failure.slotName} holds '${failure.branchName}' with a ${failure.operation ?? "operation"} in progress at ${failure.worktreePath}; cannot continue running slot gc.`,
		);
	if (failure.reason === "dirty-worktree")
		return withAction(
			entry,
			"skipped-dirty",
			`worktree has uncommitted changes at ${failure.worktreePath}`,
		);
	return withAction(entry, "error", detachFailureMessage(failure));
}

function withAction(
	entry: SlotGcEntry,
	action: SlotGcAction,
	message: string | null = null,
): SlotGcEntry {
	return { ...entry, action, message };
}

function freedSlotFromGcEntry(entry: SlotGcEntry): FreedSlot {
	return {
		slotName: entry.slotName,
		branchName: entry.branchName,
		worktreePath: entry.worktreePath,
	};
}

function gcFreeTargets(entries: readonly SlotGcEntry[]): readonly FreedSlot[] {
	return entries
		.filter((entry) => entry.action === "would-free" || entry.action === "freed")
		.map(freedSlotFromGcEntry);
}

function withCleanupBySlot(
	entries: readonly SlotGcEntry[],
	cleanup: readonly SlotFreeCleanupResult[],
): SlotGcEntry[] {
	return entries.map((entry) => ({
		...entry,
		cleanup: cleanup.filter(
			(result) => result.slotName === entry.slotName && result.branchName === entry.branchName,
		),
	}));
}

function outcomeFromEntries(entries: readonly SlotGcEntry[], isDryRun: boolean): SlotGcOutcome {
	const counts = countGcActions(entries);
	return {
		entries,
		...counts,
		dryRun: isDryRun,
		cleanupErrorCount: entries
			.flatMap((entry) => entry.cleanup)
			.filter((result) => result.status === "error").length,
	};
}

function countGcActions(entries: readonly SlotGcEntry[]): {
	freedCount: number;
	keptCount: number;
	skippedCount: number;
	errorCount: number;
} {
	let freedCount = 0;
	let keptCount = 0;
	let skippedCount = 0;
	let errorCount = 0;
	for (const entry of entries) {
		if (entry.action === "freed" || entry.action === "would-free") freedCount += 1;
		if (entry.action === "kept-open-pr" || entry.action === "kept-no-pr") keptCount += 1;
		if (entry.action === "skipped-dirty" || entry.action === "skipped-operation") skippedCount += 1;
		if (entry.action === "error") errorCount += 1;
	}
	return {
		freedCount: freedCount,
		keptCount: keptCount,
		skippedCount: skippedCount,
		errorCount: errorCount,
	};
}
